Add fixer for trailing-comma-interface rule

ismail-syed opened the pull request:
This fixer will help with integrating prettier into our formatting/linting, if we decide to go with it. 

Related to https://github.com/Shopify/tslint-config-shopify/issues/51

ismail-syed committed:
Add fixer

GoodForOneFare reviewed (changes requested):
You can add `test/rules/trailing-comma-interface/test.ts.fix`, and tslint will magically verify that your fix works ([Example fix file from the main repo]( https://github.com/palantir/tslint/blob/master/test/rules/align/members/test.ts.fix)).

Please remove the trailing comma from `moe: number,`.  This'll verify that trailing commas are enforced for last lines.

ismail-syed committed:
🔨PR fixes: Clean up test cases

ismail-syed commented:
@GoodForOneFare This is good for another quick 👀

GoodForOneFare commented on test/rules/trailing-comma-interface/test.ts.lint:
```diff
@@ -1,7 +1,7 @@
+                  ~ [0]
   larry: number;
-  ~~~~~~~~~~~~~~  [Trailing comma is required in interfaces properties.]
-  moe: number,
```
Why is `moe` deleted?  I asked for the comma to be removed from the end of its line.  This will verify that omitted commas are added by the fixer.

ismail-syed commented:
Ek, my apologies. I need to read better :/

I'll get that fixed soon. At the gym 10x'ing my fitness atm. Haha, lol.

ismail-syed committed:
Support omitted commas

ismail-syed commented:
Added back the test case without the comma. Updated the rule as well because 
```
moe: number
```
would get fixed to 
```
moe: numbe,
```

Good catch Gord, thanks! 😄

GoodForOneFare commented on src/customRules/trailingCommaInterfaceRule.ts:
```diff
@@ -4,11 +4,15 @@ import * as Lint from 'tslint';
    private static FAILURE_STRING = 'Trailing comma is required in interfaces properties.';
    public visitInterfaceDeclaration(node: ts.InterfaceDeclaration) {
+    const potentialDelimiters = [';'];
     node.members.map((propertySignature) => {
```
Nothing's consuming the result of the loop.  `map` => `forEach`.

GoodForOneFare commented on src/customRules/trailingCommaInterfaceRule.ts:
```diff
@@ -4,11 +4,15 @@ import * as Lint from 'tslint';
-                                               propertySignature.getWidth(),
-                                               TrailingCommaInterfaceWalker.FAILURE_STRING));
+        const lastChar = propertySignature.getText().slice(-1);
+        if (lastChar !== ',') {
```
Try to make your callbacks do one thing well.  Conditions are cognitive load.

Chain a couple of callbacks instead:
```
    node.members
      .filter((propertySignature) => propertySignature.getText().slice(-1) !== ',')
      .forEach((propertySignature) => {
        const lastChar = propertySignature.getText().slice(-1);
```

ismail-syed replied:
😍

GoodForOneFare commented on src/customRules/trailingCommaInterfaceRule.ts:
```diff
@@ -4,11 +4,15 @@ import * as Lint from 'tslint';
 class TrailingCommaInterfaceWalker extends Lint.RuleWalker {
    private static FAILURE_STRING = 'Trailing comma is required in interfaces properties.';
    public visitInterfaceDeclaration(node: ts.InterfaceDeclaration) {
+    const potentialDelimiters = [';'];
```
What's up with the indentation in this file?  I see 2, 3, and 4 char indents.

ismail-syed replied:
Had something wonky going on in VSCode.

GoodForOneFare commented on src/customRules/trailingCommaInterfaceRule.ts:
```diff
@@ -4,11 +4,15 @@ import * as Lint from 'tslint';
+        const lastChar = propertySignature.getText().slice(-1);
+        if (lastChar !== ',') {
+          const lastCharPosition = propertySignature.getStart() + propertySignature.getWidth() - 1;
+          const replacementText = (potentialDelimiters.indexOf(lastChar) === -1) ? `${lastChar},` : ',';
```
`potentialDelimiters` feels like an over-abstraction; it'd take a language change to add anything to this array!

Use `arr.includes(foo)` if you're not going to use the index.

Favour truthy comparisons.  Most conditions are truth-first, so it's easy to miss inversions.  eslint has a [no-negated-condition](http://eslint.org/docs/rules/no-negated-condition) rule to enforce this (I wonder if they gave it an intentionally bad name?)

So, something like:
```
const replacementText = (lastChar === ';' ? ',' : `${lastChar},`;
```

ismail-syed committed:
PR fixes 👍

GoodForOneFare reviewed (approved):
👍
